sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala

```diff
@@ -642,8 +642,13 @@ private[spark] class HiveExternalCatalog(conf: SparkConf, hadoopConf: Configurat
       if (stats.get.rowCount.isDefined) {
         statsProperties += STATISTICS_NUM_ROWS -> stats.get.rowCount.get.toString()
       }
+
+      // For datasource tables and hive serde tables created by spark 2.1 or higher,
```

Comment:
IIUC, here I think we should test when the schema read by `getRawTable` (i.e., `rawTable`) is different than the schema stored in table properties. 

Previously because two copies of schema are different, you will get the `NoSuchElement` too. That's why we should call `restoreTableMetadata` here for hive serde tables.